### ISSUE TITLE
ResolvePackageNotFound causing Binder launch to fail

### ISSUE THREAD

**Issue by emigre459** [jupyterhub/binder#149]
Every time I try to launch [my repo](https://github.com/emigre459/OSTI-API) via mybinder.org (or the auto-generated badge I put into the README file) using my conda environment.yml config file, I seem to hit a wall at step 39/44 with the error `The command '/bin/sh -c conda env update -n root -f "environment.yml" && conda clean -tipsy && conda list -n root' returned a non-zero code: 1`

It seems like it's not finding these packages, but most of them (e.g. pandas) are extremely common, so I doubt that's the issue. Any help would be much appreciated! Here is the last piece of the Build Log that includes the referenced error message:

```
Step 39/44 : RUN conda env update -n root -f "environment.yml" && conda clean -tipsy && conda list -n root
 ---> Running in ebe9a67762e4
Solving environment: ...working... failed

ResolvePackageNotFound:
  - jsonschema==2.6.0=py36hb385e00_0
  - libedit==3.1.20181209=hb402a30_0
  - tornado==5.1.1=py36h1de35cc_0
  - mistune==0.8.4=py36h1de35cc_0
  - readline==7.0=h1de35cc_5
  - markupsafe==1.1.1=py36h1de35cc_0
  - wcwidth==0.1.7=py36h8c6ec74_0
  - cryptography==2.5=py36ha12b0ac_0
  - libffi==3.2.1=h475c297_4
  - xz==5.2.4=h1de35cc_4
  - ncurses==6.1=h0a44026_1
  - python==3.6.8=haf84260_0
  - pyzmq==18.0.0=py36h0a44026_0
  - cffi==1.12.1=py36hb5b8e2f_0
  - traitlets==4.3.2=py36h65bd3ce_0
  - libsodium==1.0.16=h3efe00b_0
  - openssl==1.1.1b=h1de35cc_0
  - sqlite==3.26.0=ha441bb4_0
  - ipython_genutils==0.2.0=py36h241746c_0
  - nbformat==4.4.0=py36h827af21_0
  - tk==8.6.8=ha441bb4_0
  - yaml==0.1.7=hc338f04_2
  - zlib==1.2.11=h1de35cc_3
  - zeromq==4.3.1=h0a44026_3
  - nodejs==10.13.0=h0a44026_0
  - lxml==4.3.1=py36hef8c89e_0
  - libgfortran==3.0.1=h93005f0_2
  - libpng==1.6.36=ha441bb4_0
  - libxslt==1.1.33=h33a18ac_0
  - icu==58.2=h4b95b61_1
  - kiwisolver==1.0.1=py36h0a44026_0
  - libcxxabi==4.0.1=hcfea43d_1
  - mkl_fft==1.0.10=py36h5e564d8_0
  - scikit-learn==0.20.2=py36h27c97d8_0
  - libxml2==2.9.9=hab757c2_0
  - statsmodels==0.9.0=py36h1d22016_0
  - et_xmlfile==1.0.1=py36h1315bdc_0
  - libiconv==1.15=hdd342a3_7
  - freetype==2.9.1=hb4e5f40_0
  - numpy==1.15.4=py36hacdab7b_0
  - scipy==1.2.1=py36h1410ff5_0
  - mkl_random==1.0.2=py36h27c97d8_0
  - pyyaml==3.13=py36h1de35cc_0
  - numpy-base==1.15.4=py36h6575580_0
  - cycler==0.10.0=py36hfc81398_0
  - pandas==0.24.1=py36h0a44026_0
  - libcxx==4.0.1=hcfea43d_1
  - matplotlib==3.0.2=py36h54f8f79_0
  - appnope==0.1.0=py36hf537a9a_0

Removing intermediate container ebe9a67762e4
The command '/bin/sh -c conda env update -n root -f "environment.yml" && conda clean -tipsy && conda list -n root' returned a non-zero code: 1
```

**Comment by choldgraf** [comment]
hey @emigre459 - it looks like you created your environment.yml file using `conda`. This actually creates a set of dependencies that are *too* specific. For example, many of those `=py36xxxxx` parts are OS-dependent etc. Can you try the same list but keeping the level specificity to version numbers?

**Comment by emigre459** [comment]
Thanks so much @choldgraf! I had no idea that those were OS-specific version numbers. Upon correcting that with `conda env export --no-builds -f environment.yml`, I still get an error but now with many fewer packages causing problems:

```
ResolvePackageNotFound:
  - libcxxabi=4.0.1
  - appnope=0.1.0
  - libgfortran=3.0.1
  - libcxx=4.0.1
```

Of course, it's still an issue, and I know at least that `pandas` depends on `libcxx`, making that a pretty necessary package to load. Any idea what might be causing issues for these lingering ones?

**Comment by choldgraf** [comment]
No problem! :-)


hmmmm, my best guess would be that it's still something specific to the OS? but that's just a guess...if you don't specify version numbers, or if you just specify major versions, does that work?

**Comment by betatim** [comment]
Same thing happened here https://github.com/jupyterhub/mybinder.org-deploy/issues/906.

Packages like `appnope` only exist on OSX and so can't be installed inside a Ubuntu based docker container.

**Comment by betatim** [comment]
I added a FAQ entry to the repo2docker documentation that should answer this. It would be great if you could try to follow the guide and give feedback @emigre459.

Once that is merged we should close this issue.

**Comment by emigre459** [comment]
Sorry @betatim, I'm afraid I'm not sure which [FAQ](https://repo2docker.readthedocs.io/en/latest/faq.html) entry you're referencing. Are you perhaps referencing the [conda-freeze repo you created recently](https://github.com/binder-examples/conda-freeze)? I'm afraid I don't see how that's useful, given that it's effectively what I've already done. I would note for @choldgraf, [the other conda binder-example repo](https://github.com/binder-examples/conda) should probably have its README updated with any new guidance that comes out of this discussion (starting with the `--no-builds` flag being used as part of conda exporting). 

The other issue you referenced is definitely helpful though - are you effectively recommending that I run `jupyter-repo2docker git_hub_URL`? The steps you (@betatim) outlined in your second to last post on that issue seem to only differ from my current workflow in step 1. 

I'm going to try `jupyter-repo2docker git_hub_URL`, although now repo2docker is throwing an error saying "ModuleNotFoundError: No module named 'docker'"...

**Comment by betatim** [comment]
Sorry, the link to the proposed FAQ entry is https://github.com/jupyter/repo2docker/pull/598

**Comment by emigre459** [comment]
Ah, I see, that makes more sense now that I've seen the FAQ update suggestion (doing `jupyter-repo2docker git_hub_URL` didn't work, BTW, which is unsurprising I suppose since that's likely exactly what mybinder.org is doing anyhow). 

Do you have any idea if there's a conda or other command that can identify OS-specific packages? Or is that something that just needs to be done by hand?

**Comment by choldgraf** [comment]
That's a great question, I don't know of something that just spits out your environment but doesn't add unnecessary detail that might actually not be reproducible. I know @minrk mentioned the idea of a "jupyter-repo2docker freeze" command... Maybe that's the solution long term?

**Comment by betatim** [comment]
`jupyter-repo2docker freeze` would indeed be that command. Though it will also output a OS specific environment, but one that works on linux so repo2docker will be happy.

The FAQ entry and linked Howto guide is the manual version of `jupyter-repo2docker freeze` (till someone gets around to building that).

**Comment by emigre459** [comment]
OK, using `conda env export --no-builds -f environment.yml` and removing the packages that are still causing issues with the build (reproduced below for reference) manually from the `environment.yml` file [seemed to do the trick](https://mybinder.org/v2/gh/emigre459/OSTI-API/master). Thank you both for your help!

In case it's not specified elsewhere in the docs, I'd also note that I was using *.yaml as my file extension earlier and had to change it to *.yml for Binder. Maybe I'm uncommon in that regard, but it might be a good "gotcha" to also warn people about.

@choldgraf, I'm going to submit a pull request to modify the README for [your conda repo](https://github.com/binder-examples/conda) to reflect this (and include the packages I removed, for OSX users to have a shortcut at least), and can use the same language for your [conda-freeze repo](https://github.com/binder-examples/conda-freeze) if you like, @betatim. Just let me know if that's useful. Regardless, I'm going to close this issue since I think we've hit upon the current solution, prior to a `jupyter-repo2docker freeze` command being developed.

```
ResolvePackageNotFound:
  - libcxxabi=4.0.1
  - appnope=0.1.0
  - libgfortran=3.0.1
  - libcxx=4.0.1
```

**Comment by choldgraf** [comment]
thanks so much for the PR @emigre459 !
